GVFS/GVFS.Common/Git/LibGit2RepoInvoker.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Threading;
+
+namespace GVFS.Common.Git
+{
+    public class LibGit2RepoInvoker : IDisposable
+    {
+        private static readonly TimeSpan DefaultRepositoryDisposalPeriod = TimeSpan.FromMinutes(15);
+
+        private readonly Func<LibGit2Repo> createRepo;
+        private readonly ITracer tracer;
+        private readonly object sharedRepoLock = new object();
+        private readonly TimeSpan sharedRepositoryDisposalPeriod;
+        private volatile bool stopped;
+        private volatile int activeCallers;
+        private LibGit2Repo sharedRepo;
+        private Timer sharedRepoDisposalTimer;
+
+        public LibGit2RepoInvoker(ITracer tracer, Func<LibGit2Repo> createRepo, TimeSpan? disposalPeriod = null)
+        {
+            this.tracer = tracer;
+            this.createRepo = createRepo;
+
+            if (!disposalPeriod.HasValue || disposalPeriod.Value <= TimeSpan.Zero)
+            {
+                this.sharedRepositoryDisposalPeriod = DefaultRepositoryDisposalPeriod;
+            }
+            else
+            {
+                this.sharedRepositoryDisposalPeriod = disposalPeriod.Value;
+            }
+
+            this.sharedRepoDisposalTimer = new Timer(
+                (state) => this.DisposeSharedRepo(),
+                state: null,
+                dueTime: this.sharedRepositoryDisposalPeriod,
+                period: this.sharedRepositoryDisposalPeriod);
+        }
+
+        public bool IsActive => this.sharedRepo != null;
+
+        public void Dispose()
+        {
+            this.stopped = true;
+
+            this.sharedRepoDisposalTimer?.Dispose();
+            this.sharedRepoDisposalTimer = null;
+
+            lock (this.sharedRepoLock)
+            {
+                this.sharedRepo?.Dispose();
+                this.sharedRepo = null;
+            }
+        }
+
+        public bool TryInvoke<TResult>(Func<LibGit2Repo, TResult> function, out TResult result)
+        {
+            try
+            {
+                Interlocked.Increment(ref this.activeCallers);

```

Comment:
Would it have been an option to instead do a using statement (i.e.):
using (LibGit2Repo repo = this.createRepo()) {}
To handle the disposal?  This would mean we can have more than 1 instance of course but it ties the dispose up nicely.

Reply:
We specifically don't want multiple instances, and don't want to dispose when there are many requests coming in.

Reply:
My honest question is why don't we want multiple instances?  Is it because it will take too long?

I think a comment at the top of the file explaining the problem, and why we solve it in this way would be super helpful!

Reply:
Instead of doing the Interlocked.Increment/Decrement we could do the whole procedure in a lock.  I guess the problem here is that we want explicity want to allow multiple function runs concurrently.  Questing only being of the added complexity.

Reply:
We can't do it in a lock, because then we can't have parallel invocations of the method we are calling. It is possible to do a lock around the changes to `activeCallers`, but `Interlocked` is faster than `lock ()`.

We _want_ to allow parallel access, but we don't need independent native memory blocks to make that happen. And there is a real cost to creating and disposing the libgit2 repo with each invocation. And based on tests with #680, we break things if we have no limit to the number of native repos we create.

Reply:
@derrickstolee  that makes sense.  A think a comment on the top of the class saying something like:

Here we create  single instance of a LibGit2Repo that we attempt to clean up every 15 minutes.  It would be expensive to create/dispose a LibGit2Repo for every call so we do it on a timer schedule.  The background cleanup allows us to free memory and lingering unnecessary handles on files.